src/lib/zexe_backend/zexe_backend_common/plonk_plookup_constraint_system.ml

```diff
@@ -0,0 +1,1172 @@
+open Marlin_plonk_bindings.Types
+open Sponge
+open Unsigned.Size_t
+
+module type Gate_vector_intf = sig
+  open Unsigned
+
+  type field
+
+  type t
+
+  val create : unit -> t
+
+  val add : t -> field Plonk_plookup_gate.t -> unit
+
+  val get : t -> int -> field Plonk_plookup_gate.t
+end
+
+module Row = struct
+  type t = Public_input of int | After_public_input of int
+
+  let to_absolute ~public_input_size = function
+    | Public_input i ->
+        i
+    | After_public_input i ->
+        i + public_input_size
+end
+
+module Gate_spec = struct
+  type ('row, 'f) t =
+    { kind: Plonk_plookup_gate.Kind.t
+    ; row: 'row
+    ; lrow: 'row
+    ; lcol: Plonk_plookup_gate.Col.t
+    ; rrow: 'row
+    ; rcol: Plonk_plookup_gate.Col.t
+    ; orow: 'row
+    ; ocol: Plonk_plookup_gate.Col.t
+    ; qrow: 'row
+    ; qcol: Plonk_plookup_gate.Col.t
+    ; prow: 'row
+    ; pcol: Plonk_plookup_gate.Col.t
+    ; coeffs: 'f array }
+
+  let map_rows t ~f =
+    { t with
+      row= f t.row
+    ; lrow= f t.lrow
+    ; rrow= f t.rrow
+    ; orow= f t.orow
+    ; qrow= f t.qrow
+    ; prow= f t.prow }
+end
+
+module Hash_state = struct
+  open Core_kernel
+  module H = Digestif.SHA256
+
+  type t = H.ctx
+
+  let digest t = Md5.digest_string H.(to_raw_string (get t))
+
+  let empty = H.feed_string H.empty "plonk_constraint_system_v3"
+end
+
+(* HACK: The rust side shouldn't be using an integer for this.
+   As written, we have no correctness guarantees from either compiler.
+*)
+let col_to_index = function
+  | Plonk_plookup_gate.Col.L ->
+      0
+  | Plonk_plookup_gate.Col.R ->
+      1
+  | Plonk_plookup_gate.Col.O ->
+      2
+  | Plonk_plookup_gate.Col.Q ->
+      3
+  | Plonk_plookup_gate.Col.P ->
+      4
+
+(* HACK: The rust side shouldn't be using an integer for this.
+   As written, we have no correctness guarantees from either compiler.
+*)
+let index_to_col = function
+  | 0 ->
+      Plonk_plookup_gate.Col.L
+  | 1 ->
+      Plonk_plookup_gate.Col.R
+  | 2 ->
+      Plonk_plookup_gate.Col.O
+  | 3 ->
+      Plonk_plookup_gate.Col.Q
+  | 4 ->
+      Plonk_plookup_gate.Col.P
+  | _ ->
+      assert false
+
+module Plonk_constraint = struct
+  open Core_kernel
+
+  module T = struct
+    type ('v, 'f) t =
+      | Basic of
```

Comment:
It would be good to share most of these with the `plonk_5_wires` constraint system, and they should also be able to share the logic for interpreting them. Can you factor this out? Possibly easiest is to call directly into the 5-wires code, and just add a new dedicated gate for the `Bytes_lookup` here.